### PR TITLE
EAD_3 related updates

### PR DESCRIPTION
**Pull request by gselander**

### CODE REVIEW

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 1]
```diff
@@ -147,15 +147,17 @@ This profile also supports the alternative workflow where AS uploads the access
 
 If C has retrieved an access token, there are two options for C to upload it to RS, as further detailed in this document.
 
-1. C posts the access token to the /authz-info endpoint by using the mechanisms specified in {{Section 5.10 of RFC9200}}. If the access token is valid, RS responds to the request with a 2.01 (Created) response, after which C initiates the EDHOC protocol with RS. The communication with the /authz-info endpoint is not protected, except for the update of access rights (see {{update-access-rights-c-rs}}).
+1. C posts the access token to the /authz-info endpoint by using the mechanisms specified in {{Section 5.10 of RFC9200}}. If the access token is valid, RS responds to the request with a 2.01 (Created) response, after which C initiates the EDHOC protocol with RS. The communication with the /authz-info endpoint is typically not protected, except for the update of access rights (see {{update-access-rights-c-rs}}).
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:150]: s/RS responds to the/RS replies to the

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 2]
```diff
@@ -147,15 +147,17 @@ This profile also supports the alternative workflow where AS uploads the access
...
+After the whole procedure has completed and while the access token is valid, C can contact AS to request an update of its access rights, by sending a similar request to the /token endpoint. This request also includes a "session identifier" (see {{edhoc-parameters-object}}) provided by AS in the initial request, which allows AS to find the data it has previously shared with C. The session identifier is assigned by AS and used to identify a series of access tokens, called a "token series" (see {{token-series}}). Upon a successful update of access rights (see {{update-access-rights-c-rs}}), the new issued access token becomes the latest in its token series, but the session identifier remains the same. When the latest access token of a token series becomes invalid (e.g., when it expires or gets revoked), that token series ends.
+
+When an RS receives a request from C protected with an OSCORE Security Context derived from an EDHOC session implementing this profile, the associated session identifier, together with the authentication credential of C used in the EDHOC session, enables the RS to look up the unique access token determining the access rights of C.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:160]: s/the associated session identifier/then the associated session identifier

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 3]
```diff
@@ -269,14 +271,14 @@ This document refers to "token series" as a series of access tokens sorted in ch
...
+The general token series concept is defined in {{I-D.ietf-ace-workflow-and-params}}. In this profile, a token series is characterized by access tokens used between a given pair (C, RS) having the same session\_id in the EDHOC\_Information, see {{edhoc-parameters-object}}.
 
-AS assigns the session\_id to the EDHOC\_Information when issuing the first access token of a new series. When assigning the identifier, AS MUST ensure that it was not used in a previous series whose access tokens share the following properties with the access tokens of the new series:
+AS assigns the session\_id to the EDHOC\_Information when issuing the first access token of a new series and it remains fixed throughout the series lifetime. When assigning the identifier, AS MUST ensure that it was not used in a previous series whose access tokens share the following properties with the access tokens of the new series:
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:276]: s/AS assigns the session\_id to/AS assigns the "session\_id" to

s/series and it remains fixed/series, and that "session\_id" remains fixed

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 4]
```diff
@@ -269,14 +271,14 @@ This document refers to "token series" as a series of access tokens sorted in ch
 
 Upon a successful update of access rights, the new issued access token becomes the latest in its token series. When the latest access token of a token series becomes invalid (e.g., due to its expiration or revocation), the token series it belongs to ends.
 
-The general token series concept is defined in {{I-D.ietf-ace-workflow-and-params}}. In this profile, a token series is characterized by the access tokens associated with an EDHOC session between C and RS, and identified by the session\_id of the EDHOC\_Information, see {{edhoc-parameters-object}}. All access tokens belonging to the same token series are associated with the same session\_id, which does not change throughout the series lifetime.
+The general token series concept is defined in {{I-D.ietf-ace-workflow-and-params}}. In this profile, a token series is characterized by access tokens used between a given pair (C, RS) having the same session\_id in the EDHOC\_Information, see {{edhoc-parameters-object}}.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:274]: s/same session\_id in/same "session\_id" in

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 5]
```diff
@@ -582,30 +584,31 @@ If RS fails to find or validate AUTH_CRED_C, then RS MUST respond with an error
 
 If, instead, the access token is valid but associated with claims that RS cannot process (e.g., an unknown scope), or if any of the expected parameters is missing (e.g., any of the mandatory parameters from AS or the identifier "session\_id"), or if any parameters received in the EDHOC_Information is unrecognized, then RS MUST respond with an error response code equivalent to the CoAP code 4.00 (Bad Request). In the latter two cases, RS may provide additional information in the payload of the error response, in order to clarify what went wrong.
 
-If all validations are successful, RS MUST reply to the POST request with a 2.01 (Created) response.
+If all validations are successful, RS MUST reply to the POST request with a 2.01 (Created) response. The content of the access token is stored such that it is possible to retrieve based on session\_id and AUTH_CRED_C.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:587]: s/retrieve based on session\_id/retrieve it based on "session\_id"

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 6]
```diff
@@ -582,30 +584,31 @@ If RS fails to find or validate AUTH_CRED_C, then RS MUST respond with an error
...
-
-Editor's note: Shall we remove the EAD_1 case? The case POST /token offers already early abort, and the EAD_3 case is equally efficient but offers better protection of the access token.
+Instead of uploading the access token to the /authz-info endpoint at RS as described in {{c-rs}}, C MAY include the access token in EDHOC message\_3 by making use of the External Authorization Data field EAD_3 (see {{Section 3.8 of I-D.ietf-lake-edhoc}}), see example in {{example-with-optimization}}. In this case, the access token is encrypted between C and RS enabling protection of potential sensitive information.
 
 This document defines the EAD item EAD\_ACCESS\_TOKEN = (ead\_label, ead\_value), where:
 
-* ead\_label is the integer value TBD registered in {{iana-edhoc-ead}}, and
-* ead\_value is a CBOR byte string equal to the value of the "access_token" field of the access token response from AS (see {{as-c}}).
+* ead\_label is the integer value TBD registered in {{iana-edhoc-ead}} and
+* ead\_value is a CBOR map object with label either 0 or 1 and byte string value:
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:598]: NEW

> * ead\_value is a CBOR byte string, with value the binary serialization of a CBOR map. The CBOR map includes one field, whose label is a CBOR unsigned integer with value either 0 or 1, and whose value is a CBOR byte string. In particular:

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 7]
```diff
@@ -582,30 +584,31 @@ If RS fails to find or validate AUTH_CRED_C, then RS MUST respond with an error
...
 
-This EAD item, which may be used either in EAD\_1 or EAD\_3, is critical, i.e., it is used only with the negative value of its ead\_label, indicating that the receiving RS must either process the access token or abort the EDHOC session (see {{Section 3.8 of I-D.ietf-lake-edhoc}}).
+   * For label 0, ead\_value = { 0 : access\_token }, the map value is the CBOR byte string equal to the value of the "access\_token" field of the access token response from AS, see {{as-c}}.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:600]: OLD
> * For label 0, ead\_value = { 0 : access\_token }, the map value is the CBOR byte string equal to the value of the "access\_token" field of the access token response from AS, see {{as-c}}.

NEW
> * For label 0, ead\_value = { 0 : access\_token }, where access\_token is a CBOR byte string equal to the value of the "access\_token" field of the access token response from AS, see {{as-c}}.

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 8]
```diff
@@ -582,30 +584,31 @@ If RS fails to find or validate AUTH_CRED_C, then RS MUST respond with an error
...
+   * For label 1, ead\_value = { 1 : session\_id }, the map value is a CBOR byte string equal to the value of the "session\_id" field of the access token response from AS, see {{token-series}}.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:601]: OLD
> * For label 1, ead\_value = { 1 : session\_id }, the map value is a CBOR byte string equal to the value of the "session\_id" field of the access token response from AS, see {{token-series}}.

NEW
> * For label 1, ead\_value = { 1 : session\_id },where session\_id is a CBOR byte string equal to the value of the "session\_id" field of the access token response from AS, see {{token-series}}.

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 9]
```diff
@@ -582,30 +584,31 @@ If RS fails to find or validate AUTH_CRED_C, then RS MUST respond with an error
...
 
-Access tokens are only transported in EAD fields for the first access token of a token series and not for the update of access rights.
+This EAD item, which is used in EAD\_3, is critical, i.e., it is used only with the negative value of its ead\_label, indicating that the receiving RS must either process the access token or abort the EDHOC session (see {{Section 3.8 of I-D.ietf-lake-edhoc}}). An endpoint supporting the profile of ACE defined in this document MUST support this EAD item. When EDHOC is used with this profile, this EAD item MUST be included in EAD_3, see {{m_3}}.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:603]: s/An endpoint supporting/A node supporting

A alternative to "node" is "participant", or "Client or Resource Server". But "endpoint" is used to denote resources such as /token and /authz-info (see Section 1.1).

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 10]
```diff
@@ -582,30 +584,31 @@ If RS fails to find or validate AUTH_CRED_C, then RS MUST respond with an error
...
+
+Access tokens are only transported in EAD fields for the first access token of a token series and not for the update of access rights, see {{update-access-rights-c-rs}}.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:605]: OLD
> Access tokens are only transported in EAD fields for the first access token of a token series and not for the update of access rights, see {{update-access-rights-c-rs}}.

NEW
> Access tokens are transported in EAD_3 fields only for the first access token of a token series, and not for the update of access rights, see {{update-access-rights-c-rs}}.

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 11]
```diff
@@ -582,30 +584,31 @@ If RS fails to find or validate AUTH_CRED_C, then RS MUST respond with an error
...
 
 ## EDHOC Session and OSCORE Security Context # {#edhoc-exec}
 
-In order to mutually authenticate and establish a secure association, C and RS run the EDHOC protocol {{I-D.ietf-lake-edhoc}} with C as EDHOC Initiator and RS as EDHOC Responder.
+In order to mutually authenticate and establish secure communication for authorized access, C and RS run the EDHOC protocol {{I-D.ietf-lake-edhoc}} using the profile described in this document with C as EDHOC Initiator and RS as EDHOC Responder. When a new EDHOC session is established using this profile, any previous EDHOC session and derived security context, between the same parties associated to the access token / session_id included in EAD\_3, is removed.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:609]: s/using the profile/, using the profile

OLD
> When a new EDHOC session is established using this profile, any previous EDHOC session and derived security context, between the same parties associated to the access token / session_id included in EAD\_3, is removed.

NEW
> When a new EDHOC session S_NEW is established using this profile, it is possible that both S_NEW and another, existing session S_OLD between the same parties are associated with the same access token and with the same pair (session\_id, AUTH_CRED_C). In such a case, S_OLD and the derived OSCORE Security Context are deleted.

File: draft-ietf-ace-edhoc-oscore-profile.md [hunk 12]
```diff
@@ -643,7 +646,7 @@ Once successfully completed the EDHOC session, C and RS derive an OSCORE Securit
 
 * C and RS MUST use the OSCORE version specified in the "osc\_version" field (if present) in the EDHOC\_Information of the access token response to C, and of the access token provisioned to RS, respectively.
 
-* RS associates the derived OSCORE Security Context with the stored access token, which is bound to the authentication credential (AUTH\_CRED\_C = CRED\_I) used in the EDHOC session.
+* RS associates the derived OSCORE Security Context from the EDHOC session with its associated session\_id and authentication credential (AUTH\_CRED\_C = CRED\_I) used in the EDHOC session, which in turn identifies the access token.
```
Comment (marco-tiloca-sics) [inline draft-ietf-ace-edhoc-oscore-profile.md:649]: NEW
> RS associates the latest EDHOC session and the derived OSCORE Security Context with the stored access token, which is bound to the authentication credential (AUTH_CRED_C = CRED_I) used in the EDHOC session and with the session\_id identifying the token series that the access token belongs to.

**Comment by marco-tiloca-sics** [comment]
I've checked the commit https://github.com/ace-wg/ace-edhoc-oscore-profile/pull/4/commits/c34735adbfbda878e6309dd70a1b1741ed6b5cd8

It looks good, thanks! I see only two nits in Section 4.4:

* s/to which the access token belongs./which the access token belongs to.

* s/previous EDHOC session/previous EDHOC sessions

  For that same sentence, maybe it's better to rephrase as below altogether:

  OLD
  > When a new EDHOC session is established using this profile, previous EDHOC session associated with the same access token and the same pair (session_id, AUTH_CRED_C), and derived OSCORE Security Context are deleted, see Section 4.4.3.

  NEW
  > When a new EDHOC session is established using this profile, previous EDHOC sessions associated with the same access token and the same pair (session_id, AUTH_CRED_C) are deleted, see Section 4.4.3. The OSCORE Security Contexts derived from those EDHOC sessions are deleted.

**Comment by gselander** [comment]
Thanks Marco, I did not make the first change, there are other instances of the same construction and I think the wording is fine. For the second proposal I did another change to the paragraph. 

This should now be sufficiently ready to merge. Additional comments can be made on the main branch.

**Pull request merged**